Network specification update (#1404)
ethereum/eth2.0-specs: specs/networking/p2p-interface.md

| Name | Value | Description |
|---|---|---|

> | `REQ_RESP_MAX_SIZE` | `2**22` (4194304, 4 MiB) | The maximum size of uncompressed req/resp messages that clients will allow. |

Reviewer: After discussions about bandwidth requirements, considering dropping this down to 1MB

| Name | Value | Description |
|---|---|---|
| `GOSSIP_MAX_SIZE` | `2**20` (= 1048576, 1 MiB) | The maximum allowed size of uncompressed gossip messages. |

> | `REQ_RESP_MAX_SIZE` | `2**20` (1048576, 1 MiB) | The maximum allowed size of uncompressed req/resp chunked responses. |

Reviewer 2: Should this be renamed to `CHUNK_MAX_SIZE`?

```
request   ::= <encoding-dependent-header> | <encoded-payload>
response  ::= <response_chunk>+

> response_chunk  ::= <result> | <encoding-dependent-header> | <encoded-payload>

Reviewer 2: Is the result-per-chunk pattern deliberate? Would the result be a global result per req/rep?

result    ::= “0” | “1” | “2” | [“128” ... ”255”]
```

The encoding-dependent header may carry metadata or assertions such as the encoded payload length, for integrity and attack proofing purposes. Because req/resp streams are single-use and stream closures implicitly delimit the boundaries, it is not strictly necessary to length-prefix payloads; however, certain encodings like SSZ do, for added security.

> A `response` is formed by one or more `response_chunk`s. The exact request determines whether a response consists of a single `response_chunk` or possibly many. Responses that consist of a single SSZ-list (such as `BlocksByRange` and `BlocksByRoot`) send each list item as a `response_chunk`. All other response types (non-Lists) send a single `response_chunk`. The encoded-payload of a `response_chunk` has a maximum uncompressed byte size of `REQ_RESP_MAX_SIZE`.

Reviewer 2: This can incur in excessive fragmentation/overhead/underutilisation if list items are small, as you're effectively defining a 1:1 mapping between chunk and list element.

I'd recommend to make the boundaries of chunks full list elements, allowing a chunk to contain multiple _complete_ list elements, and disallowing partial elements or bleeding over between chunks.

It may be worth formally introducing the term `chunkable`, so you can label message fields that satisfy this property in the schemas below.

Reviewer 3: Please note that adding chunks does not increase the size of the overall payload, because we are replacing the leading 4-byte offsets that would appear in the SSZ lists before with response codes and varint lengths which will often amount to just 2 bytes per chunk.

> Clients MUST ensure the each encoded payload of a `response_chunk` is less than or equal to `REQ_RESP_MAX_SIZE`; if not, they SHOULD reset the stream immediately. Clients tracking peer reputation MAY decrement the score of the misbehaving peer under this circumstance.

Reviewer 2: Rename constant to `CHUNK_MAX_SIZE` for clarity?

#### Requesting side

Once a new stream with the protocol ID for the request type has been negotiated, the full request message SHOULD be sent immediately. The request MUST be encoded according to the encoding strategy.

The requester MUST close the write side of the stream once it finishes writing the request message. At this point, the stream will be half-closed.

> The requester MUST wait a maximum of `TTFB_TIMEOUT` for the first response byte to arrive (time to first byte—or TTFB—timeout). On that happening, the requester allows a further `RESP_TIMEOUT` for each subsequent `response_chunk` received. For responses consisting of potentially many `response_chunk`s (an SSZ-list) the requester SHOULD read from the stream until either; a) An error result is received in one of the chunks, b) The responder closes the stream,  c) More than `MAX_CHUNK_SIZE` bytes have been read for a single `response_chunk` payload or d) More than the maximum number of requested chunks are read. For requests consisting of a single `response_chunk` and a length-prefix, the requester should read the exact number of bytes defined by the length-prefix before closing the stream.

Reviewer 4: > d) More than the maximum number of requested chunks are read.

@AgeManning Should this number be specified?

Author: The two types that can request chunks are `BeaconBlocksByRange` and `BeaconBlocksByRoot`.  Each specify in the request the maximum number of blocks/chunks that should be returned. 

For `BeaconBlocksByRange` this is the `count` parameter, for `BeaconBlocksByRoot` this is the number of hashes requested.

Let me know if this doesn't answer your question :)

Reviewer 4: That makes perfect sense. Thank you for the clarification ❤️.

> The requester MUST wait a maximum of `TTFB_TIMEOUT` for the first response byte to arrive (time to first byte—or TTFB—timeout). On that happening, the requester allows a further `RESP_TIMEOUT` for each subsequent `response_chunk` received. For responses consisting of potentially many `response_chunk`s (an SSZ-list) the requester SHOULD read from the stream until either; a) An error result is received in one of the chunks, b) The responder closes the stream,  c) More than `REQ_RESP_MAX_SIZE` bytes have been read for a single `response_chunk` payload or d) More than the maximum number of requested chunks are read. For requests consisting of a single `response_chunk` and a length-prefix, the requester should read the exact number of bytes defined by the length-prefix before closing the stream.

Reviewer 2: Beware of a starvation DoS attack where attackers could deliberately trickle their chunks slowly so as to keep that socket/file descriptor busy as long as possible.

1. Use the encoding strategy to read the optional header.
2. If there are any length assertions for length `N`, it should read exactly `N` bytes from the stream, at which point an EOF should arise (no more bytes). Should this not be the case, it should be treated as a failure.
3. Deserialize the expected type, and process the request.
4. Write the response which may consist of one or more `response_chunk`s (result, optional header, payload).
5. Close their write side of the stream. At this point, the stream will be fully closed.

If steps (1), (2), or (3) fail due to invalid, malformed, or inconsistent data, the responder MUST respond in error. Clients tracking peer reputation MAY record such failures, as well as unexpected events, e.g. early stream resets.

The entire request should be read in no more than `RESP_TIMEOUT`. Upon a timeout, the responder SHOULD reset the stream.

The responder SHOULD send a `response_chunk` promptly. Chunks start with a **single-byte** response code which determines the contents of the `response_chunk` (`result` particle in the BNF grammar above).

> The response code can have one of the following values, encoded as a single unsigned byte:

Reviewer 5: We should highlight here that only the last chunk may have a value different from `success`

1. Use the encoding strategy to read the optional header.
2. If there are any length assertions for length `N`, it should read exactly `N` bytes from the stream, at which point an EOF should arise (no more bytes). Should this not be the case, it should be treated as a failure.
3. Deserialize the expected type, and process the request.
4. Write the response which may consist of one or many `response_chunks` (result, optional header, payload).
5. Close their write side of the stream. At this point, the stream will be fully closed.

If steps (1), (2), or (3) fail due to invalid, malformed, or inconsistent data, the responder MUST respond in error. Clients tracking peer reputation MAY record such failures, as well as unexpected events, e.g. early stream resets.

The entire request should be read in no more than `RESP_TIMEOUT`. Upon a timeout, the responder SHOULD reset the stream.

> The responder SHOULD send a response promptly, starting with a **single-byte** response code which determines the contents of the `response_chunk` (`result` particle in the BNF grammar above).

Author: It should be with respect to the `response_chunk`. We have a TTFB (time to first byte) which starts once the stream is open. This should probably worded more like:
```
- The responder SHOULD send a response_chunk promptly. Chunks start with a **single-byte** ...
```

```
The fields are:

- `head_fork_version`: The beacon_state `Fork` version.
- `finalized_root`: The latest finalized root the node knows about.
- `finalized_epoch`: The latest finalized epoch the node knows about.
- `head_root`: The block hash tree root corresponding to the head of the chain as seen by the sending node.
- `head_slot`: The slot corresponding to the `head_root`.

The dialing client MUST send a `Hello` request upon connection.

> This should be encoded as an SSZ-container.

Reviewer: ```suggestion
This MUST be encoded as an SSZ-container.
```

```
The fields are:

- `head_fork_version`: The beacon_state `Fork` version.
- `finalized_root`: The latest finalized root the node knows about.
- `finalized_epoch`: The latest finalized epoch the node knows about.
- `head_root`: The block hash tree root corresponding to the head of the chain as seen by the sending node.
- `head_slot`: The slot corresponding to the `head_root`.

> The dialing client MUST send a `Status` request upon connection.

Reviewer 6: Any recommendation for the scenario that the dialing client doesn't send status? Disconnect after 5 seconds? Send the dialing client a `Status`?

Reviewer 5: timeout/disconnect (see belo)

The range `[4, 127]` is RESERVED for future usage.

> This should not be encoded as an SSZ-container.

Reviewer: MUST not?
Maybe we should change this to 
```suggestion
This MUST be encoded as a single SSZ-field.
```

Requests count beacon blocks from the peer starting from `start_slot` on the chain defined by `head_block_root`. The response MUST contain no more than count blocks. `step` defines the slot increment between blocks. For example, requesting blocks starting at `start_slot` 2 with a step value of 2 would return the blocks at [2, 4, 6, …]. In cases where a slot is empty for a given slot number, no block is returned. For example, if slot 4 were empty in the previous example, the returned array would contain [2, 6, …]. A step value of 1 returns all blocks on the range `[start_slot, start_slot + count)`.

> The request is encoded as an SSZ-container, the response is not encoded as an

Reviewer: ```suggestion
The request MUST be encoded as an SSZ-container, the response MUST be encoded as a single SSZ-field.
```

> SSZ container.

Reviewer: ```suggestion
```

The request MUST be encoded as an SSZ-container.

The response MUST consist of at least one `response_chunk` and MAY consist of many. Each _successful_ `response_chunk` MUST contain a single `BeaconBlock` payload.

`BeaconBlocksByRange` is primarily used to sync historical blocks.

Clients MUST support requesting blocks since the start of the weak subjectivity period and up to the given `head_block_root`.

Clients MUST support `head_block_root` values since the latest finalized epoch.

Clients MUST respond with at least one block, if they have it.

Clients MUST order blocks by increasing slot number.

#### BeaconBlocksByRoot

**Protocol ID:** `/eth2/beacon_chain/req/beacon_blocks_by_root/1/`

> Request Content:

Reviewer 7: shouldn't both the request and response be a list of roots/blocks rather than a container ? 
Request Content:
```

(
  []HashTreeRoot
)
```
Response Content:
```
(
 []BeaconBlock
)
```

Reviewer: Looks like it

)
```

Requests blocks by their block roots. The response is a list of `BeaconBlock` whose length is less or equal to the number of requested blocks. It may be less in the case that the responding peer is missing blocks.

`BeaconBlocksByRoot` is primarily used to recover recent blocks (ex. when receiving a block or attestation whose parent is unknown).

> Both the request and the response should not be encoded as an SSZ-container.

Reviewer: ```suggestion
Both the request and the response MUST be encoded as a single SSZ-field.
```

)
```

Requests blocks by their block roots. The response is a list of `BeaconBlock` whose length is less than or equal to the number of requested blocks. It may be less in the case that the responding peer is missing blocks.

`BeaconBlocksByRoot` is primarily used to recover recent blocks (e.g. when receiving a block or attestation whose parent is unknown).

> The request MUST be encoded as an SSZ-field.

Reviewer 6: Is this still correct? The chunks are ssz containers now, `BeaconBlock`. I.e. it's a series of streamed containers rather than a ssz-field `[]BeaconBlock`.

Reviewer 6: Oops, i misread request/response. Please disregard
